src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -341,72 +343,97 @@ private void WatchForFileSystemEventsThreadStart(object arg)
+                        bool eventIsLink = (eventFlags[i] & (FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsHardlink | FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink | FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsLastHardlink)) > 0;
+
+                        if (eventIsCorrectType || ((allowDirs || allowFiles) && (eventIsLink)))
+                        {
+                            // Notify Renamed events.
+                            if (IsFlagSet(eventFlags[i], FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed))
+                            {
+                                // Find the rename that is paired to this rename, which should be the next rename in the list
+                                long pairedId = FindRenameChangePairedChange(i, eventPaths, eventFlags, eventIds);
+                                if (pairedId == long.MinValue)
+                                {
+                                    // Getting here means we have a rename without a pair, meaning it should be a create for the 
+                                    // move from unwatched folder to watcher folder scenario or a move from the watcher folder out.
+                                    // Check if the item exists on disk to check which it is
+                                    WatcherChangeTypes wct = DoesItemExist(eventPaths[i], IsFlagSet(eventFlags[i], FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile)) ?
+                                        WatcherChangeTypes.Created :
+                                        WatcherChangeTypes.Deleted;
+                                    watcher.NotifyFileSystemEventArgs(wct, relativePath);
+                                }
+                                else
                                 {
-                                    handledRenameEvents = new List<FSEventStreamEventId>();
+                                    // Remove the base directory prefix and add the paired event to the list of 
+                                    // events to skip and notify the user of the rename 
+                                    string newPathRelativeName = eventPaths[pairedId].Remove(0, _fullDirectory.Length);
+                                    watcher.NotifyRenameEventArgs(WatcherChangeTypes.Renamed, newPathRelativeName, relativePath);
+
+                                    // Create a new list, if necessary, and add the event
+                                    if (handledRenameEvents == null)
+                                    {
+                                        handledRenameEvents = new List<FSEventStreamEventId>();
+                                    }
+                                    handledRenameEvents.Add(eventIds[pairedId]);
                                 }
-                                handledRenameEvents.Add(eventIds[pairedId]);
+                                return;
```

Comment:
Nope, not correct. Another holdover from the helper method.

I'm going to switch back to the helper method so we can easily avoid doing the relative path calculation when an event hasn't passed through the filter.